File: sdk/go/iprules.go
package qeetid

import (
	"context"
	"net/http"
	"net/url"
)

type IPRule struct {
	ID          string `json:"id"`
	TenantID    string `json:"tenant_id"`
	CIDR        string `json:"cidr"`
	Action      string `json:"action"`
	Description string `json:"description,omitempty"`
	CreatedAt   string `json:"created_at"`
}

type CreateIPRuleInput struct {
	CIDR        string `json:"cidr"`
	Action      string `json:"action"`
	Description string `json:"description,omitempty"`
}

type IPRules struct{ http *httpClient }

func (r *IPRules) Create(ctx context.Context, tenantID string, in CreateIPRuleInput) (*IPRule, error) {
	var out IPRule
	err := r.http.do(ctx, http.MethodPost, "/v1/tenants/"+url.PathEscape(tenantID)+"/ip-rules", nil, in, &out, false)
	return &out, err
}

func (r *IPRules) Delete(ctx context.Context, tenantID, id string) error {
	return r.http.do(ctx, http.MethodDelete, "/v1/tenants/"+url.PathEscape(tenantID)+"/ip-rules/"+url.PathEscape(id), nil, nil, nil, true)
}

func (r *IPRules) List(ctx context.Context, tenantID string) ([]IPRule, error) {
	var env struct {
		Items []IPRule `json:"items"`
		Data  []IPRule `json:"data"`
	}
	if err := r.http.do(ctx, http.MethodGet, "/v1/tenants/"+url.PathEscape(tenantID)+"/ip-rules", nil, nil, &env, true); err != nil {
		return nil, err
	}
	if env.Items != nil {
		return env.Items, nil
	}
	return env.Data, nil
}

File: sdk/js/node/src/ipRules.ts
import type { HttpClient } from "./client.js";

export interface IpRule {
  id: string;
  tenant_id: string;
  cidr: string;
  action: string;
  description?: string;
  created_at: string;
}

export interface CreateIpRuleInput {
  cidr: string;
  action: "allow" | "deny";
  description?: string;
}

interface ListEnvelope<T> {
  items?: T[];
  data?: T[];
}

export class IpRules {
  constructor(private readonly http: HttpClient) {}

  create(tenantId: string, input: CreateIpRuleInput): Promise<IpRule> {
    return this.http.post<IpRule>(
      `/v1/tenants/${encodeURIComponent(tenantId)}/ip-rules`,
      input,
    );
  }

  delete(tenantId: string, id: string): Promise<void> {
    return this.http.delete<void>(
      `/v1/tenants/${encodeURIComponent(tenantId)}/ip-rules/${encodeURIComponent(id)}`,
    );
  }

  async list(tenantId: string): Promise<IpRule[]> {
    const res = await this.http.get<ListEnvelope<IpRule>>(
      `/v1/tenants/${encodeURIComponent(tenantId)}/ip-rules`,
    );
    return res.items ?? res.data ?? [];
  }
}
